lighthouse-core/gather/driver.js

```diff
@@ -792,6 +786,40 @@ class Driver {
     const waitForNetworkIdle = this._waitForNetworkIdle(networkQuietThresholdMs);
     // CPU listener. Resolves when the CPU has been idle for cpuQuietThresholdMs after network idle.
     let waitForCPUIdle = this._waitForNothing();
+    const cancelAll = () => {
+      waitForFCP.cancel();
+      waitForLoadEvent.cancel();
+      waitForNetworkIdle.cancel();
+      waitForCPUIdle.cancel();
+    };
+
+    // Promise that only rejects when an insecure security state is encountered
+    let securityCheckCleanup = () => {};
+    const securityCheckPromise = new Promise((_, reject) => {
```

Comment:
there's a lot to unpack in this function that needs longstanding cleanup I'll take a whack at as a followup to #6944, for now WDYT about sticking this in a `_waitForSecurityError` or something following the `{promise, cancel}` paradigm we have currently?

Reply:
The `cancelAll` and clearing the `maxTimeoutHandle` complicates that a lot. It's not clear to me how to extract into a function without using multiple parameter callbacks (or is that the cleanup you had in mind?)

Reply:
yeah this goes hand-in-hand with the `resolve` idea. I'm not sure we could easily pull it out with it still rejecting immediately

Reply:
OK, I think I see how that'd work. I'll give it a shot